README
Introduction
-------------

WebROaR is an application server that makes deployments of ruby web 
applications extremely simple. It provides an integrated solution to view 
the run time performance numbers and email notifications in case any 
exceptions occur in any of the deployed applications.

It is 5 to 55% faster than all other comparable deployment stacks for 
Ruby on Rails applications.

Check out http://webroar.in and doc/user-guide.html for more details.

Key Features
~~~~~~~~~~~~

* Maximum Performance. 
* Simplified Deployment.
* Runs Ruby on Rails(TM) as well as other Rack compliant applications. 
* Run Multiple Applications Simultaneously. 
* Implements HTTP/1.1 grammar as per RFC 2616 including support for persistent,
 and chunked requests. 
* Intelligent load balancing and dynamic reaping of stuck ruby processing
 instances. 
* Provides run time performance data for the deployed applications. 
* Generates notifications in case any exceptions occur in any of the deployed
 applications.
* Zero Downtime Application Upgrades.
* Static assets encoding.
* SSL support. 

Supported Operating Systems
---------------------------

WebROaR has been tested on Ubuntu 8.10/9.04/9.10/10.04/10.10/11.04, 
Debian 5 (64 Bit), Mac OS X (v10.5 & v10.6), Arch Linux 2009.08 and Cent OS 5.2. 
Basically, it should be able to run on any *IX OS.

Currently it does not support Microsoft(R) Windows.

Install
--------

Prerequisites 
~~~~~~~~~~~~~

- ruby 1.8.x or 1.9.x 
- ruby 1.8-dev or 1.9-dev 
- openssl ruby extension (On Ubuntu/Debian it is available as libopenssl-ruby1.X)
- zlib ruby extension
- rubygems 1.3.6 or later
- C build tools (gcc etc) (e.g. build-essential on Ubuntu, Xcode on MacOS)
- libsqlite3
- libsqlite3-dev
- gnutls 2.4.1 or later (Only necessary in case SSL support is required) 
- gnutls-dev 2.4.1 or later (Only necessary in case SSL support is required) 

Procedure
~~~~~~~~~

WebROaR can installed using the following commands:

----
sudo gem install webroar
sudo webroar install
----

[NOTE] 
.SSL Support
===============================
In case you would like to setup SSL support for the server, use the following
command line switch with *sudo webroar install* command mentioned above.
----
sudo webroar install --ssl-support
----

Configure further SSL options in the server's admin panel
  *http://<yourservername>:<port>/admin-panel/configuration*

After restarting the server, admin panel would now also be accessible over SSL.
*https://<yourservername>:<port>/admin-panel*
===============================

Or if you like living in the fast lane, you can install the edge version giving
the following commands:

----
git clone git://github.com/webroar/webroar.git
cd webroar
sudo rake install
----

[NOTE]
.SSL Support
===============================
To build with SSL support, use the following additional switch with the 
*sudo rake install* command:
----
sudo rake install ssl=yes
----

Configure further SSL options in the server's admin panel
  *http://<yourservername>:<port>/admin-panel/configuration*

After restarting the server, admin panel would now also be accessible over SSL.
*https://<yourservername>:<port>/admin-panel*
===============================



CHANGELOG
v0.7.0 - 14444-Oct-2011
--------------------

* Upgraded Admin Panel to Rails 3.0.9.
* Fixed logic to find out latest installed version.
* Fixed 'user.yml' file importing issue.

v0.6.1 - 19-Sep-2011
--------------------

New Features
~~~~~~~~~~~~

* Added Feedback and Report Bug forms in Admin Panel.
* Support to disable Admin Panel on server startup.
* Support to disable Analytics feature.
* Show backtrace and environment details of each exceptions.

Fixes
~~~~~

* Fixed server crashing issue with Ubuntu 11.04.
* Fixed file import issue by importing 'serverinternalconfig.yml' file.
* Now closing session after password change of Admin Panel.
* Fixed SSL issue in libebb.
* Changed images used in Admin Panel.
* Moved Email module to Admin Panel
* Added current directory in load_path($:) to make the server compatible with Ruby>=1.9.2.
* Fixed Javascript and Stylesheet loading problem for Rails 3 Application.
* Added helpers to run Admin Panel with Ruby 1.9.3.
* Fixed invalid byte sequence issue to run Admin Panel with Ruby 1.9.2.
* Corrected spec test cases.
* Fixed condition satement to check Rails version in adapter.
* Upgraded Admin Panel and Test App to Rails 2.3.9.
* Upgraded openflashchart_2 plugin.
* Corrected check condition to start exception tracker.
* Defined modules for server commands.
* Corrected default value of min and max workers for application
* Added link to stop an application from Admin Panel.
* Improved and cleaned up Analyzer module.
* Fixed paginate plugin issue by correcting javascript to generate Ajax request.
* Commented copyright message in Admin Panel's views.
* Hacked 'environment.rb' to fix 'version_requirements' method issue with RubyGems 1.5.x
* Fixed code compilation issue with Mac OS.

v0.5.0 - 16-Feb-2011
--------------------

New Features
~~~~~~~~~~~~

* Added support to filter out know exceptions.
* The number of exceptions displayed in exception list made configurable.
* Support gzip encoding for static assets.
* Support for adding/removing applications from its root directory with 
default parameters.

Fixes
~~~~~
* Support to render static assets for Rails 3 application.
* Updated test cases.
* Implemented new YAML parser to handle lists and associative arrays.
* Added client side validations on Admin Panel.
* Code clean-up in Admin Panel.
* Corrected rendering of graph partials.
* Improve popup windows used to display help messages.
* Corrected checkbox selection functionality on exception listing page.
* Check Cache-Control header before choosing encoding. This fixes an issue of 
static content encoding on IE6.
* Update values on Admin Panel home page using single Ajax request.
* Wrapped multi statement macros in do..while structure.
* Fixed rack dependency issue.
* Improve the logic of exiting worker gracefully if there is an error.
* Fixed starling dependency issue with Rails 3 application.
* Kept strict dependency for Rspec 1.3.x.
* Fixed QUERY_STRING issue faced with Rails 3 application.
* Clean-up and improve migration scripts.
* Updated libyaml to v0.1.3
* Upgraded Libev to v3.9.
* Improved Ajax call on Analytics page.


v0.4.0 - 09-Aug-2010
--------------------

New Features
~~~~~~~~~~~~

* Supports deflate encoding for static assets.
* Support for passing environment variables to application.
* Support to build server with static Ruby library.

Fixes
~~~~~

* Tested on Ruby 1.9.2 RC2
* Tested on Ubuntu 10.04.
* Directory path expanded when application is added from command line. (It
would allow to pass relative path)
* Fixed incorrect response code of static assets by using 'timegm' function 
instead of 'mktime'.
* Requiring version.rb using relative path, to work with Ruby 1.9.2.
* Added Config::CONFIG['rubyhdrdir'] in ruby headers search path for Ruby 1.9.
* Sends '503 Service Unavailable' response, if pending message queue overflows.
* Code cleanup in Rake tasks.
* Enhanced SCGI request parsing in Admin Panel to handle chunked request.
* Added tests for application add, remove and zero down time during restart.
* Added Config::CONFIG['includedir'] in include directory list.
* Supports Rails 3 application.
* Created 'config.yml' file with default configuration if it is not found while
importing.
* Added support to start/stop/restart static-worker.
* Support to view analytics data for all the application, earlier it was only 
for the application which has analytics enabled.
* Saving log files after executing load test.
* Improved rake task to execute test-suite.
* Set Expires headers from Admin Panel.
* Server specification code cleanup in Admin Panel.
* Passing application configuration parameters to worker using SCGI.
* Email notification for the captured exceptions made configurable.
* Made many server parameters configurable through config file 
'server_internal_config.yml'.
* Sends ERROR control message, if application could not be loaded by worker.
* Wait for 30 minutes before creating new worker, if three consecutive newly 
created worker got timed out.
* Removed 'Cache-Control' response header from static assets. Its incorrect 
handling choking IE 6.


v0.3.1 - 22-Mar-2010
--------------------

* Refactored and improved the code for handling restarts of an application.
* Corrected output emitted for version command.
* Fixed the incorrect server version being sent in the HTTP Response Headers.
* Updated description of the gem.
* Minor updates in the README and User Guide.


v0.3.0 - 19-Mar-2010
--------------------

New Features
~~~~~~~~~~~~

* Support for Zero Downtime Application Upgrades.
* Full support for adding/removing applications from command line. (Earlier 
it was only possible through the Admin Panel.)
* Full support for non interactive installation of the server.
* Support for passing additional include and library paths for server 
compilation.
* Display error trace directly on the Admin Panel or the console if the 
application start/stop/restart command fails.

Fixes
~~~~~


src/head/wr_config.h
/* WebROaR - Ruby Application Server - http://webroar.in/
 * Copyright (C) 2009  Goonj LLC
 *
 * This file is part of WebROaR.
 *
 * WebROaR is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * WebROaR is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with WebROaR.  If not, see <http://www.gnu.org/licenses/>.
 */
#ifndef WR_CONFIG_H_
#define WR_CONFIG_H_

// Default max_worker
#define WR_MAX_WKR 8
// Default min_worker
#define WR_MIN_WKR 4
// Default request listining port
#define WR_DEFAULT_SVR_PORT 3000

// Maximum value allowed for number of workers
#define WR_ALLOWED_MAX_WORKERS 20

#define WR_MAX_PENDING_WKR 10
#define WR_MAX_ADD_TIMEOUT_COUNTER 3
#define WR_WKR_ADD_WAIT_TIME 1800.

#ifdef HAVE_GNUTLS
// Default SSL listining port
#define WR_DEFAULT_SSL_PORT 443
#endif

// Default application environment
#define WR_DEFAULT_ENV "production"
// Admin Panel base uri
#define WR_ADMIN_PANEL_BASE_URL "/admin-panel"
// Admin Panel base uri
#define WR_ADMIN_PANEL_APP_NAME "Admin Panel"
// Parameter value to enable application analytics
#define WR_ANALYTICS_ON "enabled"

// Appliction to serve static files
#define WR_STATIC_FILE_SERVER_NAME "static-worker"
// Minimum number of worker to serve static content
#define WR_STATIC_SVR_MIN_WKR 4
// Maximum number of worker to serve static content
#define WR_STATIC_SVR_MAX_WKR 8
#define WR_HTTP_FILE_PATH        "FILE_PATH"
#define WR_HTTP_FILE_PATH_LEN      9

#define RESP_CODE               "RESP_CODE"
#define RESP_CONTENT_LENGTH   "RESP_CONTENT_LEN"
#define SCGI_CONTENT_LENGTH "CONTENT_LENGTH"

/** Folder and file related macros */
// Server control sock path
#define WR_CTL_SOCK_PATH "/tmp/webroar_controller_sock"
// 'webroar-worker' request listining sock path
#define WR_WKR_SOCK_PATH "/tmp/webroar_worker_sock"
// Server log file name
#define WR_LOG_FILE "webroar.log"
// Server configuration file name
#define WR_CONF_FILE "config.yml"
// Path seperator
#define WR_PATH_SEPARATOR "/"
// Ruby library folder  name
#define WR_RUBY_LIB_DIR "ruby_lib"
// Admin Panel folder name
#define WR_ADMIN_PANEL_DIR "admin_panel"
// Source folder name
#define WR_SRC_DIR "src"
// Executable directory name
#define WR_BIN_DIR "bin"
// File to store Server control port or sock path
#define WR_TMP_SOCK_FILE  "/tmp/webroar.sock"
// File to store process pid having high resident memory
#define WR_HIGH_RSS_PID_FILE  "/var/run/high_rss.pid"
// Executable file name used to create workers
#define WR_WKR_BIN "webroar-worker"
// File to store Server process id
#define WR_PID_FILE  "/var/run/webroar.pid"
#define WR_SERVER "WebROaR"
#define WR_VERSION "0.3.2"
#define WR_MIME_TYPE_PATH "/conf/mime_type.yml"
#define WR_CONF_PATH "/conf/config.yml"
#define WR_SERVER_INTERNAL_CONF_PATH "/conf/server_internal_config.yml"

/** Private macros */
/** Application level macros */
// Pending request queue size
#define WR_MSG_QUE_SIZE 2048
// Ratio of pending request and active worker required to add worker
#define WR_MAX_REQ_RATIO 1
// Ratio of pending request and active worker required to remove worker
#define WR_MIN_REQ_RATIO 3
// Number of seconds with ratio higher than WR_MAX_REQ_RATIO to add worker.
#define WR_HIGH_LOAD_LIMIT 2.
// Number of seconds with ratio lower than WR_MIN_REQ_RATIO to remove worker.
#define WR_LOW_LOAD_LIMIT 600.
// Number of seconds to wait for add signal from worker after incrementing the active worker count
#define WR_WKR_ADD_TIMEOUT 25.
// Number of seconds to wait for response before killing worker.
#define WR_WKR_KILL_TIMEOUT 10.

// Number of seconds worker can remain idle, before sending PING.
#define WR_WKR_IDLE_TIME 60.
// Number of seconds head should wait for PING reply.
#define WR_PING_WAIT_TIME 15.
// Number of time head should try sending PING before killing worker.
#define WR_PING_TRIALS 2

//Number of seconds head would wait before closing the connection for kepp-alive request
#define WR_SVR_KEEP_ALIVE       1
#define WR_SVR_KEEP_ALIVE_TIME   15.

// Request body more than this size sholud store in file
#define WR_REQ_BODY_MAX_SIZE 1024*64
// Bufffer size to store request headers in 'webroar-worker'
#define WR_BUF_SIZE (1024 * 10)

#define WR_TINY_STR_LEN            16
#define WR_SHORT_STR_LEN          32
#define WR_STR_LEN                64
#define WR_LONG_STR_LEN            128
#define WR_LONG_LONG_STR_LEN      512
#define WR_FILE_PATH_LEN          1024

#define WR_DEFAULT_PREFIX_HASH    5381
#define WR_MSG_SIZE              1024

#define WR_REQ_CONN_POOL         10
#define WR_CTL_CONN_POOL        5

/** Request size limitation **/
#define WR_MAX_REQ_URI_LEN            12288    /* 12*1024 */
#define WR_MAX_REQ_PATH_LEN          1024
#define WR_MAX_REQ_FRAG_LEN      1024
#define WR_MAX_REQ_QRY_STR_LEN      10240    /* 10*1024 */
#define WR_MAX_REQ_FIELD_NAME_LEN      256      /*  */
#define WR_MAX_REQ_FIELD_VALUE_LEN      81920    /* 80*1024 */
#define WR_MAX_REQ_HEADER_LEN        112640  /* (1024 * (80 + 32)) */
#define WR_MAX_REQ_HEADER_NO          40

#define WR_STACKTRACE_SIZE 50

#ifdef L_DEBUG
#define WR_CONN_ID      "CONNECTION_ID"
#define WR_CONN_ID_LEN    19
#define WR_REQ_ID        "REQ_ID"
#define WR_REQ_ID_LEN      12
#endif

#define WR_MAX_HOST_NAMES 16

#define TRUE  1
#define FALSE 0

#endif /*WR_CONFIG_H_*/
